In [1]:
!pip freeze > requirements.txt

In [2]:
import pandas

In [3]:
df = pandas.read_csv('https://raw.githubusercontent.com/Uwanga14/netflix_customer_churn/refs/heads/main/preprocessed.csv')

In [4]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,payment_method_Debit Card,payment_method_Gift Card,payment_method_PayPal,favorite_genre_Action,favorite_genre_Comedy,favorite_genre_Documentary,favorite_genre_Drama,favorite_genre_Horror,favorite_genre_Romance,favorite_genre_Sci-Fi
0,51,14,29,8,1,1,0,0,0,1,...,0,1,0,1,0,0,0,0,0,0
1,47,0,19,13,1,5,0,0,0,1,...,0,1,0,0,0,0,0,0,0,1
2,27,16,10,13,0,2,1,1,0,0,...,0,0,0,0,0,0,1,0,0,0
3,53,4,12,17,1,2,0,0,0,1,...,0,0,0,0,0,0,0,1,0,0
4,56,1,13,13,1,2,0,0,0,1,...,0,0,0,1,0,0,0,0,0,0


In [5]:
# 1. Combined viewing engagement
df['engagement_score'] = (
    df['watch_hours'] * df['avg_watch_time_per_day']
)

In [6]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,payment_method_Gift Card,payment_method_PayPal,favorite_genre_Action,favorite_genre_Comedy,favorite_genre_Documentary,favorite_genre_Drama,favorite_genre_Horror,favorite_genre_Romance,favorite_genre_Sci-Fi,engagement_score
0,51,14,29,8,1,1,0,0,0,1,...,1,0,1,0,0,0,0,0,0,0
1,47,0,19,13,1,5,0,0,0,1,...,1,0,0,0,0,0,0,0,1,0
2,27,16,10,13,0,2,1,1,0,0,...,0,0,0,0,0,1,0,0,0,16
3,53,4,12,17,1,2,0,0,0,1,...,0,0,0,0,0,0,1,0,0,0
4,56,1,13,13,1,2,0,0,0,1,...,0,0,1,0,0,0,0,0,0,0


In [7]:
# 2. Viewing intensity relative to inactivity
df['watch_intensity'] = (
    df['watch_hours'] / (df['last_login_days'] + 1)
).astype(int)

In [8]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,payment_method_PayPal,favorite_genre_Action,favorite_genre_Comedy,favorite_genre_Documentary,favorite_genre_Drama,favorite_genre_Horror,favorite_genre_Romance,favorite_genre_Sci-Fi,engagement_score,watch_intensity
0,51,14,29,8,1,1,0,0,0,1,...,0,1,0,0,0,0,0,0,0,0
1,47,0,19,13,1,5,0,0,0,1,...,0,0,0,0,0,0,0,1,0,0
2,27,16,10,13,0,2,1,1,0,0,...,0,0,0,0,1,0,0,0,16,1
3,53,4,12,17,1,2,0,0,0,1,...,0,0,0,0,0,1,0,0,0,0
4,56,1,13,13,1,2,0,0,0,1,...,0,1,0,0,0,0,0,0,0,0


In [9]:
df.watch_intensity.value_counts()

watch_intensity
0     4102
1      459
2      169
3       76
4       45
5       36
6       22
7       17
11      10
9        9
8        8
10       7
14       7
12       5
13       4
16       3
18       3
21       2
15       2
17       2
32       2
22       1
25       1
52       1
98       1
20       1
34       1
28       1
27       1
29       1
26       1
Name: count, dtype: int64

In [10]:
# 3. Identify inactive customers
df['inactive_customer'] = (
    df['last_login_days'] >= 14
).astype(int)

In [11]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,favorite_genre_Action,favorite_genre_Comedy,favorite_genre_Documentary,favorite_genre_Drama,favorite_genre_Horror,favorite_genre_Romance,favorite_genre_Sci-Fi,engagement_score,watch_intensity,inactive_customer
0,51,14,29,8,1,1,0,0,0,1,...,1,0,0,0,0,0,0,0,0,1
1,47,0,19,13,1,5,0,0,0,1,...,0,0,0,0,0,0,1,0,0,1
2,27,16,10,13,0,2,1,1,0,0,...,0,0,0,1,0,0,0,16,1,0
3,53,4,12,17,1,2,0,0,0,1,...,0,0,0,0,1,0,0,0,0,0
4,56,1,13,13,1,2,0,0,0,1,...,1,0,0,0,0,0,0,0,0,0


0-Active
1-Inactive

In [12]:
df.inactive_customer.value_counts()

inactive_customer
1    3851
0    1149
Name: count, dtype: int64

In [13]:
# 4. Identify accounts with multiple profiles
df['multiple_profiles'] = (
    df['number_of_profiles'] > 1
).astype(int)

In [14]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,favorite_genre_Comedy,favorite_genre_Documentary,favorite_genre_Drama,favorite_genre_Horror,favorite_genre_Romance,favorite_genre_Sci-Fi,engagement_score,watch_intensity,inactive_customer,multiple_profiles
0,51,14,29,8,1,1,0,0,0,1,...,0,0,0,0,0,0,0,0,1,0
1,47,0,19,13,1,5,0,0,0,1,...,0,0,0,0,0,1,0,0,1,1
2,27,16,10,13,0,2,1,1,0,0,...,0,0,1,0,0,0,16,1,0,1
3,53,4,12,17,1,2,0,0,0,1,...,0,0,0,1,0,0,0,0,0,1
4,56,1,13,13,1,2,0,0,0,1,...,0,0,0,0,0,0,0,0,0,1


In [15]:
# Display the new features
df[
    [
        'watch_hours',
        'avg_watch_time_per_day',
        'last_login_days',
        'number_of_profiles',
        'engagement_score',
        'watch_intensity',
        'inactive_customer',
        'multiple_profiles'
    ]
].head()

,watch_hours,avg_watch_time_per_day,last_login_days,number_of_profiles,engagement_score,watch_intensity,inactive_customer,multiple_profiles
0,14,0,29,1,0,0,1,0
1,0,0,19,5,0,0,1,1
2,16,1,10,2,16,1,0,1
3,4,0,12,2,0,0,0,1
4,1,0,13,2,0,0,0,1


In [16]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 40 columns):
 #   Column                      Non-Null Count  Dtype
---  ------                      --------------  -----
 0   age                         5000 non-null   int64
 1   watch_hours                 5000 non-null   int64
 2   last_login_days             5000 non-null   int64
 3   monthly_fee                 5000 non-null   int64
 4   churned                     5000 non-null   int64
 5   number_of_profiles          5000 non-null   int64
 6   avg_watch_time_per_day      5000 non-null   int64
 7   gender_0                    5000 non-null   int64
 8   gender_1                    5000 non-null   int64
 9   gender_2                    5000 non-null   int64
 10  subscription_type_Basic     5000 non-null   int64
 11  subscription_type_Premium   5000 non-null   int64
 12  subscription_type_Standard  5000 non-null   int64
 13  region_Africa               5000 non-null   int64
 14  region_Asia        

In [17]:
print(
    df.groupby('inactive_customer')['churned'].mean()
)

inactive_customer
0    0.190601
1    0.596209
Name: churned, dtype: float64


In [18]:
#Recency Score
df['recency_score'] = 1 / (df['last_login_days'] + 1).round(2)

Customers who logged in recently are less likely to churn.

In [19]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,favorite_genre_Documentary,favorite_genre_Drama,favorite_genre_Horror,favorite_genre_Romance,favorite_genre_Sci-Fi,engagement_score,watch_intensity,inactive_customer,multiple_profiles,recency_score
0,51,14,29,8,1,1,0,0,0,1,...,0,0,0,0,0,0,0,1,0,0.033333
1,47,0,19,13,1,5,0,0,0,1,...,0,0,0,0,1,0,0,1,1,0.050000
2,27,16,10,13,0,2,1,1,0,0,...,0,1,0,0,0,16,1,0,1,0.090909
3,53,4,12,17,1,2,0,0,0,1,...,0,0,1,0,0,0,0,0,1,0.076923
4,56,1,13,13,1,2,0,0,0,1,...,0,0,0,0,0,0,0,0,1,0.071429


In [20]:
print(
    df.groupby('subscription_type_Premium')['churned'].mean()
)

subscription_type_Premium
0    0.536740
1    0.437094
Name: churned, dtype: float64


In [21]:
#Premium Subscriber
df['premium_subscriber'] = (
df['subscription_type_Premium']
).astype(int)

In [22]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,favorite_genre_Drama,favorite_genre_Horror,favorite_genre_Romance,favorite_genre_Sci-Fi,engagement_score,watch_intensity,inactive_customer,multiple_profiles,recency_score,premium_subscriber
0,51,14,29,8,1,1,0,0,0,1,...,0,0,0,0,0,0,1,0,0.033333,0
1,47,0,19,13,1,5,0,0,0,1,...,0,0,0,1,0,0,1,1,0.050000,0
2,27,16,10,13,0,2,1,1,0,0,...,1,0,0,0,16,1,0,1,0.090909,0
3,53,4,12,17,1,2,0,0,0,1,...,0,1,0,0,0,0,0,1,0.076923,1
4,56,1,13,13,1,2,0,0,0,1,...,0,0,0,0,0,0,0,1,0.071429,0


In [23]:
#Low Engagement
df['low_engagement'] = (
    df['watch_hours'] < 5
).astype(int)


In [24]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,favorite_genre_Horror,favorite_genre_Romance,favorite_genre_Sci-Fi,engagement_score,watch_intensity,inactive_customer,multiple_profiles,recency_score,premium_subscriber,low_engagement
0,51,14,29,8,1,1,0,0,0,1,...,0,0,0,0,0,1,0,0.033333,0,0
1,47,0,19,13,1,5,0,0,0,1,...,0,0,1,0,0,1,1,0.050000,0,1
2,27,16,10,13,0,2,1,1,0,0,...,0,0,0,16,1,0,1,0.090909,0,0
3,53,4,12,17,1,2,0,0,0,1,...,1,0,0,0,0,0,1,0.076923,1,1
4,56,1,13,13,1,2,0,0,0,1,...,0,0,0,0,0,0,1,0.071429,0,1


In [25]:
value_counts_low_engagement = df['low_engagement'].value_counts()

In [26]:
value_counts_low_engagement

low_engagement
0    3232
1    1768
Name: count, dtype: int64

0- No
1- Yes
Accounts with low engagement are less than those who engage.
Very low viewing activity often indicates churn risk.

In [27]:
#High Engagement
df['high_engagement'] = (
df['watch_hours'] > 20
).astype(int)

In [28]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,favorite_genre_Romance,favorite_genre_Sci-Fi,engagement_score,watch_intensity,inactive_customer,multiple_profiles,recency_score,premium_subscriber,low_engagement,high_engagement
0,51,14,29,8,1,1,0,0,0,1,...,0,0,0,0,1,0,0.033333,0,0,0
1,47,0,19,13,1,5,0,0,0,1,...,0,1,0,0,1,1,0.050000,0,1,0
2,27,16,10,13,0,2,1,1,0,0,...,0,0,16,1,0,1,0.090909,0,0,0
3,53,4,12,17,1,2,0,0,0,1,...,0,0,0,0,0,1,0.076923,1,1,0
4,56,1,13,13,1,2,0,0,0,1,...,0,0,0,0,0,1,0.071429,0,1,0


In [29]:
value_counts_high_engagement = df['high_engagement'].value_counts()
value_counts_high_engagement

high_engagement
0    4188
1     812
Name: count, dtype: int64

0- No
1- Yes
Highly engaged users are typically less likely to cancel.

In [30]:
df['shared_account'] = (
    df['number_of_profiles'] >= 3
).astype(int)

In [31]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,favorite_genre_Sci-Fi,engagement_score,watch_intensity,inactive_customer,multiple_profiles,recency_score,premium_subscriber,low_engagement,high_engagement,shared_account
0,51,14,29,8,1,1,0,0,0,1,...,0,0,0,1,0,0.033333,0,0,0,0
1,47,0,19,13,1,5,0,0,0,1,...,1,0,0,1,1,0.050000,0,1,0,1
2,27,16,10,13,0,2,1,1,0,0,...,0,16,1,0,1,0.090909,0,0,0,0
3,53,4,12,17,1,2,0,0,0,1,...,0,0,0,0,1,0.076923,1,1,0,0
4,56,1,13,13,1,2,0,0,0,1,...,0,0,0,0,1,0.071429,0,1,0,0


Accounts shared by multiple users may be more valuable and less likely to churn.

In [32]:
df['daily_engagement_ratio'] = (
df['avg_watch_time_per_day'] /
(df['watch_hours'] + 1)
).round(2)

In [33]:
df.head(5)

,age,watch_hours,last_login_days,monthly_fee,churned,number_of_profiles,avg_watch_time_per_day,gender_0,gender_1,gender_2,...,engagement_score,watch_intensity,inactive_customer,multiple_profiles,recency_score,premium_subscriber,low_engagement,high_engagement,shared_account,daily_engagement_ratio
0,51,14,29,8,1,1,0,0,0,1,...,0,0,1,0,0.033333,0,0,0,0,0.00
1,47,0,19,13,1,5,0,0,0,1,...,0,0,1,1,0.050000,0,1,0,1,0.00
2,27,16,10,13,0,2,1,1,0,0,...,16,1,0,1,0.090909,0,0,0,0,0.06
3,53,4,12,17,1,2,0,0,0,1,...,0,0,0,1,0.076923,1,1,0,0,0.00
4,56,1,13,13,1,2,0,0,0,1,...,0,0,0,1,0.071429,0,1,0,0,0.00


In [34]:
feature_corr = df[
    [
        'watch_hours',
        'avg_watch_time_per_day',
        'last_login_days',
        'number_of_profiles',
        'engagement_score',
        'watch_intensity',
        'inactive_customer',
        'multiple_profiles',
        'churned'
    ]
].corr()['churned'].sort_values(ascending=False)

print(feature_corr)

churned                   1.000000
last_login_days           0.471590
inactive_customer         0.341288
multiple_profiles        -0.081959
engagement_score         -0.102443
number_of_profiles       -0.158614
watch_intensity          -0.217059
avg_watch_time_per_day   -0.217654
watch_hours              -0.480526
Name: churned, dtype: float64


Customers who have not logged in recently and are classified as inactive show the highest likelihood of churn, with last_login_days (0.472) and inactive_customer (0.341) being the strongest positive predictors.

Higher platform engagement reduces churn risk, with watch_hours (-0.481), avg_watch_time_per_day (-0.218), and watch_intensity (-0.217) showing that more active viewers are less likely to cancel their subscriptions.

In [35]:
df.to_csv('featureEngineering.csv', index=False)